Add expand-all action in search

okmttdhr opened the pull request:
<!-- Thank you for submitting a Pull Request. Please: 
* Read our Pull Request guidelines:
  https://github.com/Microsoft/vscode/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `master` branch.
* Include a description of the proposed changes and how to test them. 
-->

This PR fixes https://github.com/microsoft/vscode/issues/28090

<!--
[![Image from Gyazo](https://i.gyazo.com/dac0f2557bcd1a8194c193d7dcf7edde.gif)](https://gyazo.com/dac0f2557bcd1a8194c193d7dcf7edde)
-->

[![Image from Gyazo](https://i.gyazo.com/060e65b6ab90c22a6f1595a10c9d3ec4.gif)](https://gyazo.com/060e65b6ab90c22a6f1595a10c9d3ec4)


(Please let me know if there's a smart way to implement unit tests).

okmttdhr commented on src/vs/workbench/contrib/search/browser/searchActions.ts:
```diff
@@ -363,6 +363,60 @@ export class CollapseDeepestExpandedLevelAction extends Action {
+		@IViewletService private readonly viewletService: IViewletService,
+		@IPanelService private readonly panelService: IPanelService
+	) {
+		super(id, label, 'search-action codicon-chrome-restore');
```
How can I add a new `codicon` for `expand-all` ?
(`codicon-chrome-restore` is just a placeholder)

JacksonKearl replied:
@misolori Can we get `expand-all` codicon? Probably just `collapse-all` but with a plus.

JacksonKearl commented:
Thanks for taking this on! As discussed in the original issue, I think it would be better if this were only visible when there are items to expand. The trick is the three states:

- Everything collapsed => Show "Expand All"
- Everything expanded => Show "Collapse All"
- Otherwise => ???

As the default behavior for search is now to show all results expanded, I think we should lean towards providing Collapse functionality, and the "Expand All" action should only appear if everything is collapsed, with only the "Collapse All" being visible otherwise.

It might even make sense to only add a context menu entry for expanding all, as with #84087 the search results will almost always be fully expanded

cc @roblourens for ideas

roblourens commented:
I agree, would rather not see two buttons at the same time. Just having a context menu option would be easier...

okmttdhr commented:
Thank you for your feedback!

I'll try to hide/show each button according to items.

I'm thinking about adding property like `this.visible` to `ExpandAllAction` and `CollapseDeepestExpandedLevelAction`, but do you have any ideas??

okmttdhr commented:
Hello @JacksonKearl (CC: @roblourens )
I updated this PR, could you please take a look at it?

JacksonKearl commented:
![Dec-12-2019 19-07-28](https://user-images.githubusercontent.com/8586769/70766474-aef30380-1d12-11ea-973a-8475f7d8e990.gif)

I'm noticing some unexpected behavior, where in a multiroot workspace if I expend the first node then run the action, everything expands. But if I expand any other node, the all collapse.

JacksonKearl commented:
I fixed the issue with the first node being treated differently (the while-loop was skipping the first node, changed to do-while) and added a hook to update the action when the tree state is toggled outside of this action. Should be good to go once we get an icon. Thanks for your work on this!

misolori commented:
@roblourens @JacksonKearl I updated Codicons in master to now include a `expand-all` icon:

![image](https://user-images.githubusercontent.com/35271042/70834087-3fcaed00-1dae-11ea-8e8b-b859f418bb95.png)

okmttdhr commented:
> I fixed the issue with the first node being treated differently (the while-loop was skipping the first node, changed to do-while) and added a hook to update the action when the tree state is toggled outside of this action. Should be good to go once we get an icon. Thanks for your work on this!

Thank you for your fixing and merging this PR 🙇✨☺️
